### PR TITLE
Update EIP-3540: Restrict EOF version in create

### PR DESCRIPTION
**Pull request by chfast**
Forbid EOF → legacy contract creation.

### CODE REVIEW

File: EIPS/eip-3540.md [hunk 1]
```diff
@@ -135,8 +135,10 @@ For clarity, the *container* refers to the complete account code, while *code* r
 
 For clarity, the *EOF prefix* together with a version number *n* is denoted as the *EOFn prefix*, e.g. *EOF1 prefix*.
 
-1. If *initcode's container* has EOF1 prefix it must be valid EOF1 code.
-2. If *code's container* has EOF1 prefix it must be valid EOF1 code.
+1. If *initcode's container* has EOF1 prefix it MUST be valid EOF1 code.
+2. If *code's container* has EOF1 prefix it MUST be valid EOF1 code.
+3. If *initcode's container* is valid EOF1 code the result *code's container* MUST be valid EOF1 code (i.e. EOF1 initcode MUST NOT produce legacy code).
```
[the remark this answers is not in the document]
Comment (gumb0) [reply EIPS/eip-3540.md:140]: (Maybe something like "returning empty code is still allowed")

File: EIPS/eip-3540.md [hunk 2]
```diff
@@ -166,18 +168,11 @@ minimal set of restrictions related to the contract creation: if *initcode* or *
...
-- EOF1 contract can execute `CREATE` instruction with legacy *initcode* to create new legacy contract,
-- Legacy contract can execute `CREATE` instruction with EOF1 *initcode* to create new EOF1 contract,
-- Legacy contract can execute `CREATE` instruction with EOF1 *initcode* to create new legacy contract,
-- etc.
-
-To limit the number of exotic bytecode version combinations, additional restrictions are considered, but currently are not part of the specification:
-
-1. The EOF version of *initcode* must much the version of *code*.
-2. An EOF1 contract must not create legacy contracts.
+This guarantees than a cluster of EOF contracts will never spawn new legacy contracts.
```
Comment (gumb0) [inline EIPS/eip-3540.md:174]:
```suggestion
This guarantees that a cluster of EOF contracts will never spawn new legacy contracts.
```

File: EIPS/eip-3540.md [hunk 3]
```diff
@@ -166,18 +168,11 @@ minimal set of restrictions related to the contract creation: if *initcode* or *
...
+Furthermore, so exotic contract creation combinations are eliminated (e.g. EOF1 contract creating new EOF1 contract with legacy *initcode*).
```
Comment (gumb0) [inline EIPS/eip-3540.md:175]: some?
```suggestion
Furthermore, some exotic contract creation combinations are eliminated (e.g. EOF1 contract creating new EOF1 contract with legacy *initcode*).
```

File: EIPS/eip-3540.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (chfast) [reply EIPS/eip-3540.md:140]: What does returning empty do?

[the remark this answers is not in the document]
Comment (gumb0) [reply EIPS/eip-3540.md:140]: I think it creates account with empty code.

[the remark this answers is not in the document]
Comment (lightclient) [reply EIPS/eip-3540.md:140]: Since empty isn't valid EOF code, I think we should disallow.

**Review by lightclient — APPROVED** [review]

File: EIPS/eip-3540.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lightclient) [reply EIPS/eip-3540.md:140]: Can we just put a note that empty code is invalid? cc @gumb0

File: EIPS/eip-3540.md (hunk 2 above)
Comment (axic) [reply EIPS/eip-3540.md:174]: Added.

File: EIPS/eip-3540.md (hunk 3 above)
Comment (axic) [reply EIPS/eip-3540.md:175]: Added.

File: EIPS/eip-3540.md [hunk 4]
```diff
@@ -138,7 +138,7 @@ For clarity, the *EOF prefix* together with a version number *n* is denoted as t
 
 1. If *initcode's container* has EOF1 prefix it MUST be valid EOF1 code.
 2. If *code's container* has EOF1 prefix it MUST be valid EOF1 code.
-3. If *initcode's container* is valid EOF1 code the result *code's container* MUST be valid EOF1 code (i.e. EOF1 initcode MUST NOT produce legacy code).
+3. If *initcode's container* is valid EOF1 code the result *code's container* MUST be valid EOF1 code, in particular MUST NOT be empty (i.e. EOF1 initcode MUST NOT produce legacy code).
```
Comment (axic) [inline EIPS/eip-3540.md:141]:
```suggestion
3. If *initcode's container* is valid EOF1 code the resulting *code's container* MUST be valid EOF1 code (i.e. it MUST NOT be empty and MUST NOT produce legacy code).
```

**Review by axic — APPROVED** [review]

**Pull request merged**